src-exe/cabal-plan.hs

```diff
@@ -868,14 +870,27 @@ trPairs :: Tr a -> [(Int,a,a)]
 trPairs (No _ i js) =
     [ (n, i, j) | No n j _ <- js ] ++ concatMap trPairs js
 
+-- Compute a set of elements reachable from the given root in a graph.
+reachableFrom :: Ord a => a -> Map a (Set a) -> Set a
+reachableFrom x graph =
```

Comment:
https://hackage.haskell.org/package/topograph-1/docs/Topograph.html#v:closure + https://hackage.haskell.org/package/topograph-1/docs/Topograph.html#v:adjacencyMap

Let you precalculate that.

Reply:
nice!